File: src/precalculus_sum_difference_formulas_medium.py
"""
precalculus - sum_difference_formulas (medium)
"""

from problem_utils import *

SPECIAL = [
    (15,  45, 30, '-'),
    (75,  45, 30, '+'),
    (105, 60, 45, '+'),
    (120, 90, 30, '+'),
    (135, 90, 45, '+'),
    (150, 90, 60, '+'),
    (165, 120, 45, '+'),
    (195, 150, 45, '+'),
    (210, 180, 30, '+'),
    (225, 180, 45, '+'),
    (240, 180, 60, '+'),
    (255, 210, 45, '+'),
    (285, 270, 15, '+'),
    (300, 270, 30, '+'),
    (315, 270, 45, '+'),
    (330, 300, 30, '+'),
]


def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # Evaluate sin/cos/tan of sum/difference of standard angles
        a1 = choice([30, 45, 60, 90, 120, 135, 150])
        a2 = choice([30, 45, 60])
        is_sum = choice([True, False])
        func = choice(['sin', 'cos', 'tan'])

        result_deg = (a1 + a2) if is_sum else (a1 - a2)
        result_rad = result_deg * pi / 180
        op = '+' if is_sum else '-'

        if func == 'sin':
            ans = simplify(sin(result_rad))
            formula = r"\sin(A\pm B)=\sin A\cos B\pm\cos A\sin B"
        elif func == 'cos':
            ans = simplify(cos(result_rad))
            formula = r"\cos(A\pm B)=\cos A\cos B\mp\sin A\sin B"
        else:
            # Avoid undefined tan
            if result_deg % 180 == 90:
                func = 'sin'
                ans = simplify(sin(result_rad))
                formula = r"\sin(A\pm B)=\sin A\cos B\pm\cos A\sin B"
            else:
                try:
                    ans = simplify(tan(result_rad))
                    formula = r"\tan(A\pm B)=\frac{\tan A\pm\tan B}{1\mp\tan A\tan B}"
                except Exception:
                    func = 'cos'
                    ans = simplify(cos(result_rad))
                    formula = r"\cos(A\pm B)=\cos A\cos B\mp\sin A\sin B"

        return problem(
            question=f"Evaluate $\\{func}({a1}^\\circ {op} {a2}^\\circ)$ exactly.",
            answer=ans,
            difficulty=(1300, 1400),
            topic="precalculus/sum_difference_formulas",
            solution=steps(
                f"Use: ${formula}$",
                f"Substitute $A = {a1}^\\circ$, $B = {a2}^\\circ$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # Simplify with variables: sin(ax)/cos(bx) combo → single function
        a = choice([2, 3, 4, 5, 6])
        b = choice([1, 2, 3, 4])
        while a == b:
            b = choice([1, 2, 3, 4])

        variant = choice(['expand_sin', 'expand_cos', 'collapse_sin', 'collapse_cos'])

        if variant == 'expand_sin':
            angle_add = choice([pi/6, pi/4, pi/3, pi/2])
            lhs = sin(x + angle_add)
            ans = expand_trig(lhs)
            return problem(
                question=f"Expand $\\sin\\!\\left(x + {latex(angle_add)}\\right)$ using the sum formula.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/sum_difference_formulas",
                solution=steps(
                    "$\\sin(A+B) = \\sin A\\cos B + \\cos A\\sin B$",
                    f"$= \\sin x\\cos{latex(angle_add)} + \\cos x\\sin{latex(angle_add)}$",
                    f"$= {latex(ans)}$"
                )
            )

        elif variant == 'expand_cos':
            angle_add = choice([pi/6, pi/4, pi/3, pi/2])
            is_sum_v = choice([True, False])
            if is_sum_v:
                lhs = cos(x + angle_add)
            else:
                lhs = cos(x - angle_add)
            ans = expand_trig(lhs)
            pm = '+' if is_sum_v else '-'
            return problem(
                question=f"Expand $\\cos\\!\\left(x {pm} {latex(angle_add)}\\right)$ using the {'sum' if is_sum_v else 'difference'} formula.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/sum_difference_formulas",
                solution=steps(
                    f"$\\cos(A{pm}B) = \\cos A\\cos B {'\\mp' if is_sum_v else '\\pm'}\\sin A\\sin B$",
                    f"$= {latex(ans)}$"
                )
            )

        elif variant == 'collapse_sin':
            # sin(ax)cos(bx) - cos(ax)sin(bx) = sin((a-b)x)
            ans = sin((a - b) * x)
            return problem(
                question=f"Simplify $\\sin({a}x)\\cos({b}x) - \\cos({a}x)\\sin({b}x)$.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/sum_difference_formulas",
                solution=steps(
                    "$\\sin(A-B)=\\sin A\\cos B - \\cos A\\sin B$",
                    f"$= \\sin({a}x - {b}x) = \\sin({a-b}x)$"
                )
            )

        else:
            # cos(ax)cos(bx) + sin(ax)sin(bx) = cos((a-b)x)
            ans = cos((a - b) * x)
            return problem(
                question=f"Simplify $\\cos({a}x)\\cos({b}x) + \\sin({a}x)\\sin({b}x)$.",
                answer=ans,
                difficulty=(1400, 1500),
                topic="precalculus/sum_difference_formulas",
                solution=steps(
                    "$\\cos(A-B)=\\cos A\\cos B + \\sin A\\sin B$",
                    f"$= \\cos({a}x - {b}x) = \\cos({a-b}x)$"
                )
            )

    elif problem_type == 3:
        # tan addition / subtraction formula
        a1 = choice([15, 30, 45, 60, 75, 120, 135, 150])
        a2 = choice([30, 45, 60])
        is_sum = choice([True, False])
        result_deg = (a1 + a2) if is_sum else (a1 - a2)

        if result_deg % 180 == 90:
            # undefined — skip to safe values
            a1, a2, is_sum = 15, 45, False  # 15-45 = -30 — tan(-30) well defined
            result_deg = -30

        result_rad = result_deg * pi / 180
        ans = simplify(tan(result_rad))
        op = '+' if is_sum else '-'
        formula = r"\tan(A\pm B) = \frac{\tan A \pm \tan B}{1 \mp \tan A\tan B}"

        return problem(
            question=f"Evaluate $\\tan({a1}^\\circ {op} {a2}^\\circ)$ using the {'sum' if is_sum else 'difference'} formula.",
            answer=ans,
            difficulty=(1400, 1500),
            topic="precalculus/sum_difference_formulas",
            solution=steps(
                f"Formula: ${formula}$",
                f"$\\tan({a1}^\\circ) = {latex(simplify(tan(a1*pi/180)))}$, "
                f"$\\tan({a2}^\\circ) = {latex(simplify(tan(a2*pi/180)))}$",
                f"$= {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 4:
        # Exact value of non-standard angle — wider target pool
        target, a1, a2, op = choice(SPECIAL)
        func = choice(['sin', 'cos', 'tan'])

        angle_rad = target * pi / 180
        if func == 'tan' and target % 180 == 90:
            func = 'cos'

        if func == 'sin':
            ans = simplify(sin(angle_rad))
        elif func == 'cos':
            ans = simplify(cos(angle_rad))
        else:
            try:
                ans = simplify(tan(angle_rad))
                if not ans.is_finite:
                    func = 'sin'
                    ans = simplify(sin(angle_rad))
            except Exception:
                func = 'cos'
                ans = simplify(cos(angle_rad))

        return problem(
            question=f"Find the exact value of $\\{func}({target}^\\circ)$ using a sum or difference formula.",
            answer=ans,
            difficulty=(1500, 1600),
            topic="precalculus/sum_difference_formulas",
            solution=steps(
                f"${target}^\\circ = {a1}^\\circ {op} {a2}^\\circ$",
                f"Apply the $\\{func}$ {'sum' if op == '+' else 'difference'} formula",
                f"$= {latex(ans)}$"
            )
        )

    else:
        # Solve sin(x + c) = val or cos(x - c) = val on [0, 2pi)
        angle_const = choice([pi/6, pi/4, pi/3, pi/2, 2*pi/3])
        target_val = choice([Rational(1, 2), sqrt(2)/2, sqrt(3)/2, 0, -Rational(1,2)])
        is_sin = choice([True, False])
        pm = choice(['+', '-'])

        if is_sin:
            if pm == '+':
                equation = Eq(sin(x + angle_const), target_val)
                question = f"Solve for $x \\in [0, 2\\pi)$: $\\sin\\!\\left(x + {latex(angle_const)}\\right) = {latex(target_val)}$"
            else:
                equation = Eq(sin(x - angle_const), target_val)
                question = f"Solve for $x \\in [0, 2\\pi)$: $\\sin\\!\\left(x - {latex(angle_const)}\\right) = {latex(target_val)}$"
        else:
            if pm == '+':
                equation = Eq(cos(x + angle_const), target_val)
                question = f"Solve for $x \\in [0, 2\\pi)$: $\\cos\\!\\left(x + {latex(angle_const)}\\right) = {latex(target_val)}$"
            else:
                equation = Eq(cos(x - angle_const), target_val)
                question = f"Solve for $x \\in [0, 2\\pi)$: $\\cos\\!\\left(x - {latex(angle_const)}\\right) = {latex(target_val)}$"

        sols_full = solve(equation, x)
        sols = [s for s in sols_full if s.is_real and 0 <= s < 2*pi]
        # also check by substituting 2pi-shifted values
        extended = []
        for s in sols:
            for k in range(-1, 3):
                cand = s + 2*pi*k
                if 0 <= cand < 2*pi:
                    extended.append(cand)
        if extended:
            ans = FiniteSet(*[simplify(v) for v in set(extended)])
        else:
            ans = FiniteSet()

        return problem(
            question=question,
            answer=ans,
            difficulty=(1500, 1600),
            topic="precalculus/sum_difference_formulas",
            solution=steps(
                f"Let $u = x {pm} {latex(angle_const)}$, solve $\\{'sin' if is_sin else 'cos'}(u) = {latex(target_val)}$",
                f"Find solutions for $u$, then subtract the constant to get $x$",
                f"Solutions in $[0, 2\\pi)$: ${latex(ans)}$"
            )
        )

emit(generate())

File: src/precalculus_sum_difference_formulas_medium.jl
# precalculus - sum_difference_formulas (medium)
# Generated: 2026-03-08T20:30:27.314294

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("precalculus/sum_difference_formulas")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Evaluate sin(A+B) given sin(A), cos(A), sin(B), cos(B)
        # Choose angles in first quadrant for simplicity
        sin_a_num = randint(3, 12)
        sin_a_den = randint(sin_a_num + 1, 15)
        cos_a = sqrt(sin_a_den^2 - sin_a_num^2) // sin_a_den
        
        sin_b_num = randint(4, 11)
        sin_b_den = randint(sin_b_num + 1, 16)
        cos_b = sqrt(sin_b_den^2 - sin_b_num^2) // sin_b_den
        
        sin_a = sin_a_num // sin_a_den
        sin_b = sin_b_num // sin_b_den
        
        ans = sin_a * cos_b + cos_a * sin_b
        
        problem(
            question="If \\(\\sin(A) = $(tex(sin_a))\\) and \\(\\cos(A) = $(tex(cos_a))\\), and \\(\\sin(B) = $(tex(sin_b))\\) and \\(\\cos(B) = $(tex(cos_b))\\), find \\(\\sin(A + B)\\).",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                "Use the sum formula: \\(\\sin(A + B) = \\sin(A)\\cos(B) + \\cos(A)\\sin(B)\\)",
                sol("Substitute", sin_a * cos_b + cos_a * sin_b),
                sol("Simplify", ans)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Evaluate cos(A-B) given sin and cos values
        cos_a_num = randint(5, 12)
        cos_a_den = randint(cos_a_num + 1, 15)
        sin_a = sqrt(cos_a_den^2 - cos_a_num^2) // cos_a_den
        cos_a = cos_a_num // cos_a_den
        
        cos_b_num = randint(6, 11)
        cos_b_den = randint(cos_b_num + 1, 16)
        sin_b = sqrt(cos_b_den^2 - cos_b_num^2) // cos_b_den
        cos_b = cos_b_num // cos_b_den
        
        ans = cos_a * cos_b + sin_a * sin_b
        
        problem(
            question="If \\(\\sin(A) = $(tex(sin_a))\\), \\(\\cos(A) = $(tex(cos_a))\\), \\(\\sin(B) = $(tex(sin_b))\\), and \\(\\cos(B) = $(tex(cos_b))\\), find \\(\\cos(A - B)\\).",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                "Use the difference formula: \\(\\cos(A - B) = \\cos(A)\\cos(B) + \\sin(A)\\sin(B)\\)",
                sol("Substitute", cos_a * cos_b + sin_a * sin_b),
                sol("Simplify", ans)
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Exact value using sum/difference: sin(75°) = sin(45° + 30°)
        angles = [
            (75, 45, 30, "sin", "+"),
            (15, 45, 30, "sin", "-"),
            (75, 45, 30, "cos", "-"),
            (105, 60, 45, "sin", "+"),
            (15, 60, 45, "cos", "+")
        ]
        
        angle, a1, a2, func, op = choice(angles)
        
        if func == "sin"
            if op == "+"
                if angle == 75
                    ans_expr = "\\frac{\\sqrt{6} + \\sqrt{2}}{4}"
                else  # 105
                    ans_expr = "\\frac{\\sqrt{6} + \\sqrt{2}}{4}"
                end
            else  # op == "-"
                ans_expr = "\\frac{\\sqrt{6} - \\sqrt{2}}{4}"
            end
        else  # cos
            if op == "+"
                ans_expr = "\\frac{\\sqrt{6} - \\sqrt{2}}{4}"
            else
                ans_expr = "\\frac{\\sqrt{6} + \\sqrt{2}}{4}"
            end
        end
        
        problem(
            question="Find the exact value of \\(\\\$(func)($(angle)°)\\) using sum or difference formulas.",
            answer=ans_expr,
            difficulty=(1400, 1600),
            answer_type="expression",
            solution=steps(
                "Write \\($(angle)° = $(a1)° $(op) $(a2)°\\)",
                "Apply the $(func == "sin" ? "sine" : "cosine") $(op == "+" ? "sum" : "difference") formula",
                "Substitute known values and simplify to get \\($(ans_expr)\\)"
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Simplify using sum/difference formula
        a = randint(2, 8) * 15
        b = randint(1, 6) * 15
        
        if rand(Bool)
            # sin(a)cos(b) + cos(a)sin(b)
            ans = a + b
            problem(
                question="Simplify using sum or difference formulas: \\(\\sin($(a)°)\\cos($(b)°) + \\cos($(a)°)\\sin($(b)°)\\)",
                answer="\\sin($(ans)°)",
                difficulty=(1300, 1500),
                answer_type="expression",
                solution=steps(
                    "Recognize the sine sum formula: \\(\\sin(A + B) = \\sin(A)\\cos(B) + \\cos(A)\\sin(B)\\)",
                    "Here \\(A = $(a)°\\) and \\(B = $(b)°\\)",
                    sol("Answer", "\\sin($(ans)°)")
                ),
                time=90
            )
        else
            # cos(a)cos(b) - sin(a)sin(b)
            ans = a + b
            problem(
                question="Simplify using sum or difference formulas: \\(\\cos($(a)°)\\cos($(b)°) - \\sin($(a)°)\\sin($(b)°)\\)",
                answer="\\cos($(ans)°)",
                difficulty=(1300, 1500),
                answer_type="expression",
                solution=steps(
                    "Recognize the cosine sum formula: \\(\\cos(A + B) = \\cos(A)\\cos(B) - \\sin(A)\\sin(B)\\)",
                    "Here \\(A = $(a)°\\) and \\(B = $(b)°\\)",
                    sol("Answer", "\\cos($(ans)°)")
                ),
                time=90
            )
        end
        
    elseif problem_type == 5
        # tan(A+B) or tan(A-B) formula
        tan_a = randint(2, 8)
        tan_b = randint(2, 7)
        
        if rand(Bool)
            # tan(A+B)
            ans_num = tan_a + tan_b
            ans_den = 1 - tan_a * tan_b
            
            if ans_den != 0
                ans = ans_num // ans_den
                problem(
                    question="If \\(\\tan(A) = $(tan_a)\\) and \\(\\tan(B) = $(tan_b)\\), find \\(\\tan(A + B)\\).",
                    answer=ans,
                    difficulty=(1300, 1600),
                    solution=steps(
                        "Use the tangent sum formula: \\(\\tan(A + B) = \\frac{\\tan(A) + \\tan(B)}{1 - \\tan(A)\\tan(B)}\\)",
                        sol("Substitute", "\\frac{$(tan_a) + $(tan_b)}{1 - $(tan_a) \\cdot $(tan_b)} = \\frac{$(ans_num)}{$(ans_den)}"),
                        sol("Answer", ans)
                    ),
                    time=120
                )
            else
                ans_num = tan_a - tan_b
                ans_den = 1 + tan_a * tan_b
                ans = ans_num // ans_den
                problem(
                    question="If \\(\\tan(A) = $(tan_a)\\) and \\(\\tan(B) = $(tan_b)\\), find \\(\\tan(A - B)\\).",
                    answer=ans,
                    difficulty=(1300, 1600),
                    solution=steps(
                        "Use the tangent difference formula: \\(\\tan(A - B) = \\frac{\\tan(A) - \\tan(B)}{1 + \\tan(A)\\tan(B)}\\)",
                        sol("Substitute", "\\frac{$(tan_a) - $(tan_b)}{1 + $(tan_a) \\cdot $(tan_b)} = \\frac{$(ans_num)}{$(ans_den)}"),
                        sol("Answer", ans)
                    ),
                    time=120
                )
            end
        else
            # tan(A-B)
            ans_num = tan_a - tan_b
            ans_den = 1 + tan_a * tan_b
            ans = ans_num // ans_den
            
            problem(
                question="If \\(\\tan(A) = $(tan_a)\\) and \\(\\tan(B) = $(tan_b)\\), find \\(\\tan(A - B)\\).",
                answer=ans,
                difficulty=(1300, 1600),
                solution=steps(
                    "Use the tangent difference formula: \\(\\tan(A - B) = \\frac{\\tan(A) - \\tan(B)}{1 + \\tan(A)\\tan(B)}\\)",
                    sol("Substitute", "\\frac{$(tan_a) - $(tan_b)}{1 + $(tan_a) \\cdot $(tan_b)} = \\frac{$(ans_num)}{$(ans_den)}"),
                    sol("Answer", ans)
                ),
                time=120
            )
        end
        
    else  # problem_type == 6
        # Verify identity or prove using sum/difference formulas
        coeff = choice([2, 3, 4])
        
        if rand(Bool)
            problem(
                question="Use sum formulas to prove that \\(\\sin(x + \\pi) = -\\sin(x)\\).",
                answer="Proven",
                difficulty=(1500, 1700),
                answer_type="proof",
                solution=steps(
                    "Apply sine sum formula: \\(\\sin(x + \\pi) = \\sin(x)\\cos(\\pi) + \\cos(x)\\sin(\\pi)\\)",
                    "Substitute \\(\\cos(\\pi) = -1\\) and \\(\\sin(\\pi) = 0\\)",
                    "Simplify: \\(\\sin(x)(-1) + \\cos(x)(0) = -\\sin(x)\\)"
                ),
                time=150
            )
        else
            problem(
                question="Use sum formulas to prove that \\(\\cos(x + \\frac{\\pi}{2}) = -\\sin(x)\\).",
                answer="Proven",
                difficulty=(1500, 1700),
                answer_type="proof",
                solution=steps(
                    "Apply cosine sum formula: \\(\\cos(x + \\frac{\\pi}{2}) = \\cos(x)\\cos(\\frac{\\pi}{2}) - \\sin(x)\\sin(\\frac{\\pi}{2})\\)",
                    "Substitute \\(\\cos(\\frac{\\pi}{2}) = 0\\) and \\(\\sin(\\frac{\\pi}{2}) = 1\\)",
                    "Simplify: \\(\\cos(x)(0) - \\sin(x)(1) = -\\sin(x)\\)"
                ),
                time=150
            )
        end
    end
end